Project1

MIZINCHIK opened the pull request.

MIZINCHIK committed:
Implement Dictionary interface & SetDictionary class implementing the former

Dictionary can add & remove words. SetDictionary is based on a set provided in a constructor.

MIZINCHIK committed:
Fix unused attribute random

MIZINCHIK committed:
Implement game model

Model reacts to provided input (single chars) and gives State on input. State contains the current word mask that may be shown to the player as well as the status of their last guess.

MIZINCHIK committed:
Implement util classes for reading from terminal and logging standard messages

MIZINCHIK committed:
Implement controller

Controller handles user input and output via util classes and control the game flow with via the Model.

MIZINCHIK committed:
Implement an entry point for an example instance of the game

MIZINCHIK committed:
Test the main logic of the game flow

MIZINCHIK committed:
Fix codestyle

infomouse commented on src/main/java/edu/project1/Dictionary/Dictionary.java:
```diff
@@ -0,0 +1,7 @@
+package edu.project1.Dictionary;
```
Пакеты надо называть в нижнем регистре.
https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html

infomouse commented on src/main/java/edu/project1/Util/TerminalWriter.java:
```diff
@@ -0,0 +1,89 @@
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class TerminalWriter {
```
Идея вынести представление в отдельный класс правильная, но реализовывать надо тоже через интерфейс.
Тогда можно будет легко заменить терминал на гуи.

infomouse commented on src/main/java/edu/project1/Controller.java:
```diff
@@ -0,0 +1,89 @@
+                continue;
+            }
+            State state = model.guess(userInput.charAt(0));
+            switch (state.result()) {
```
Такая логика должна быть в классе, отвечающим за представление. Как показать пользователю текущее состояние игры - задача представления.

infomouse commented on src/main/java/edu/project1/Model/Model.java:
```diff
@@ -0,0 +1,82 @@
+        if (!correctGuess) {
+            mistakes++;
+        }
+        State result = new State(new String(userWordMask), correctGuess
```
В конструкторе не надо делать тернарников. Это не читаемо. Лучше вынести в переменную.

infomouse commented on src/main/java/edu/project1/Model/Model.java:
```diff
@@ -0,0 +1,82 @@
+        return true;
+    }
+
+    public String getUserWordMask() {
```
По-хорошему не должно быть таких геттеров. Все общение должно быть через класс State, потому что он точно отражает текущее состояние модели. Геттером же можно получить какое-то промежуточное значение одного поля, при этом сама модель может находится в не консистентном состоянии.

MIZINCHIK replied:
Теперь метод используется только при получении маски свежевыданного слова.

MIZINCHIK committed:
Fix package naming

All symbols to lower case.

MIZINCHIK committed:
Refactor view

Extract to its own package. Make an interface for extensibility. Rename methods to more universal names. Move logic of showing current state to view.

MIZINCHIK committed:
Add missing attributes of state

MIZINCHIK committed:
Model: Delete useless getters & Refactor State creation

MIZINCHIK committed:
Controller: Fit to new view logic and state encapsualtion

MIZINCHIK committed:
Fix codestyle
